### PR TITLE
PhpdocToCommentFixer - Ignore annotations

### PR DESCRIPTION
**Pull request by localheinz**
This PR

* [x] adds a failing test case
* [x] adjusts `PhpDocToComment` fixer to ignore annotations

💁‍♂ For reference, see https://github.com/vimeo/psalm/issues/2344.

/cc @theseer

closes https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/4446

### CODE REVIEW

**Comment by localheinz** [comment]
@kubawerlos 

> Can we go with option `ignored_tags` like for [`CommentToPhpdocFixer`](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4385/files) instead?

Makes a lot of sense, will adjust!

**Comment by Taluu** [comment]
I don't think adding just an option to "ignore some tags" is really dx-friendly. There are A LOT of annotations tags (doctrine, psalm, php ea inspections, ... and so on).

I think that if a line in a comment starts with `* @`, it should be safe to consider that it's an annot...

**Comment by kubawerlos** [comment]
> I think that if a line in a comment starts with `* @`, it should be safe to consider that it's an annot...

@Taluu it seems you are totally missing the point of [PHPDoc](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#3-definitions).

The first criteria if a comment is/should be PHPDoc is: if it's placed before the structural element or not, so relying on the content is not a solution.

> There are A LOT of annotations tags (doctrine, psalm, php ea inspections, ... and so on).

Doctrine's tags must be in PHPDocs, so this is not really good example.

**Comment by Taluu** [comment]
> The DocBlock is the combination of a DocComment, which is a type of comment, and a PHPDoc entity.

DocBlocks (which are also annotations, as PHPDocs tags actually) shouldn't be transformed then.

**Comment by kubawerlos** [comment]
Why so? If they are not placed before the structural element they should.

**Comment by Taluu** [comment]
It's not because they're not placed above a structural element that they shouldn't be an annotation.

Think of `@var`, `@codeCoverageIgnore`, ... and so on. `@var` is a valid phpdoc tag, yet not always used on a structural element...

**Comment by kubawerlos** [comment]
>  `@var` is a valid phpdoc tag, yet not always used on a structural element...

Like when? I'd say it always is.

**Comment by theseer** [comment]
> > `@var` is a valid phpdoc tag, yet not always used on a structural element...
> 
> Like when? I'd say it always is.

Like at any inline location where a type declaration might be needed to get IDE autocomplete or have psalm annotations:

```php
/** @var $foo SomeType */
$foo = $obj->someMethod();
```

Not saying a "proper" fix would probably be to have correct return type on the method `someMethod` - but that may not be an option.

Or, scroll back to the top and read the referenced issue at psalm (https://github.com/vimeo/psalm/issues/2344) and the very reason why @localheinz created the PR.

**Comment by theseer** [comment]
Actually, I think there are various - potentially arbitrary - reasons to have inline docblocks with annotations: To disable inspections, to add assertions for or otherwise configure static analyzers, code coverage generation, ...

Having `PhpDocToCommentFixer` unconditionally replace inline docblocks breaks all of them. It makes perfect sense to replace mere comments that just happend to use the "wrong" syntax of course, but a docblock regardless of its location with any type of annotation in it is more than just a comment.

**Comment by kubawerlos** [comment]
Adding `/** @var $foo SomeType */` before variable is perfectly fine usage of PHPDoc.

> To disable inspections, to add assertions for or otherwise configure static analyzers, code coverage generation, ...

All these reasons seem to have defined a set of annotations that would be uses -> thus the idea of having a list of tags to ignore.

> Having `PhpDocToCommentFixer` unconditionally replace inline docblocks breaks all of them.

Not unconditionally, simply according to the PSR5, which clarifies what PHPDoc is and what isn't.

If your proposition is to ignore PSR5 and trying to guess if PHPDoc should be converted comment basing of its content I think you need to create an issue to discuss it there - not here in a pull request.

File: src/Tokenizer/Analyzer/CommentsAnalyzer.php [hunk 1]
```diff
@@ -92,6 +92,11 @@ public function isBeforeStructuralElement(Tokens $tokens, $index)
         return false;
     }
 
+    public function isAnnotation(Token $token)
+    {
+        return 1 === Preg::match('#\/\*{2}\s+.*@\w+#m', $token->getContent());
```
Comment (SpacePossum) [inline src/Tokenizer/Analyzer/CommentsAnalyzer.php]: shouldn't this also do a type check? For example this string literal `$a = "/** @var int $foo */"` and HEREDOC/NOWDOC s etc., should all not be considered annotations I think?

**Comment by julienfalque** [comment]
I've got this issue a few times and I admit it is quite annoying. Adding a whitelist option would keep current stricly PSR-5 compliant behavior but I think it's not very DX-friendly and I prefer ignoring any `/**` comment that starts with a tag.

**Comment by SpacePossum** [comment]
I like the the proposal of @kubawerlos , but I can see why it might be a bit of trouble to configure the exceptions. To be clear `@var` already works fine AFAIK.

The cases I found are those from popular SCA's like `@psalm-suppress`  (but i prefer baselines file now).

I propose to add the ignore option in the next minor release. On the next major we release with a new default value for it containing `@psalm-suppress` and some other that people can provide here that are biting them.

**Comment by GrahamCampbell** [comment]
For some background, the reason this fixer changes some phpdoc containing annotations to comments is because of the proposed PSR phpdoc spec, which says that doc comments must document structural elements, so the sort of doc comments you have in your tests are illegal according to that spec.

HOWEVER, things have changed since that spec was first drafted, and the role of phpdoc has changed. No longer is phpdoc just for documentation generators to show the public api of classes, but these days we have genuinely usable static analyzers, and indeed these tools are now dictating the syntax (and semantics) of phpdoc (and annotations), and the sensible locations for phpdoc within a piece of code has changed, to allow for, for example, the specification of the type of an internal variable: something that was not relevant, even in 2014 (since the tooling was just not there, back then).

With all this in mind, I think this is a good change, in principle. The outstanding review comments need to be addressed, as do the failing tests, then this PR should be ready for 2nd reviews (and merge, if the php-cs-fixer team are happy!). ;)

**Comment by GrahamCampbell** [comment]
Unrelated, but phpstorm processes comments as phpdoc, in error. Kinda annoying, but I guess they do it because:

1. not everyone understands the difference between a comment and a doc comment
2. people have run php-cs-fixer on their code and it's changed their doc comments to comments :trollface:

File: src/Tokenizer/Analyzer/CommentsAnalyzer.php [hunk 2]
```diff
@@ -92,6 +92,11 @@ public function isBeforeStructuralElement(Tokens $tokens, $index)
         return false;
     }
 
+    public function isAnnotation(Token $token)
```
Comment (julienfalque) [inline src/Tokenizer/Analyzer/CommentsAnalyzer.php:95]: Shouldn't it accepts the `Tokens` list and an index like other methods do?

File: tests/Tokenizer/Analyzer/CommentsAnalyzerTest.php [hunk 3]
```diff
@@ -302,4 +302,121 @@ public function providePhpdocCandidatePhp74Cases()
...
+$foo = 900;
+',
+            ],
+            [
+                '<?php
+
+/* @psalm-suppress hmmm */
+#foo = 900;
+',
+            ],
```
Comment (julienfalque) [inline tests/Tokenizer/Analyzer/CommentsAnalyzerTest.php:419]: Maybe we can add cases like e.g. `/** no "at" sign */` and `/** @ followed by a space */`?

File: src/Fixer/Phpdoc/PhpdocToCommentFixer.php [hunk 4]
```diff
@@ -88,6 +88,10 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
                 continue;
             }
 
+            if ($commentsAnalyzer->isAnnotation($token)) {
```
Comment (kubawerlos) [inline src/Fixer/Phpdoc/PhpdocToCommentFixer.php:91]: Can we have this configurable to maintain BC? An option `ignore_annotations` (with default to `false`) seems the way.

**Review by kubawerlos — CHANGES_REQUESTED** [review]

**Comment by VincentLanglet** [comment]
What's the status of this ?
I started https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5572 before discovering this PR.

The fact that `/** @psalm- */` annotation are modified is really annoying.

**Comment by masi** [comment]
Please, merge this code. Having @psalm-suppress (and possible many other linter/checker annotations) as collateral damage is no fun.

**Comment by kubawerlos** [comment]
@masi as you can see there is not agreement in the discussion - you can always disable the rule if does any harm to your code.

**Comment by masi** [comment]
But the rule is fine for all comments unless they contain (certain) annotations.
Of course if Psalm would allow it's annotations within standard comments I would also be happy :)

**Comment by kubawerlos** [comment]
How is it fine? In the current state it does not convert to comment if there is any annotation in PHPDoc which makes the description "Docblocks should only be used on structural elements" not being true.

I'd prefer this one to go through: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5572

**Comment by masi** [comment]
I don't care about what the PHPDoc . I'm using annotations for my IDE and other tools. Eg I'm using Psalm and this tool insists on docblocs for its annotations.

For sake of consistency I prefer to use a single * for normal comments.

Perhaps I failed to see the subtleties. If #5572 is in, I'm happy.

Sorry, for any misunderstandings and the added confusion.

**Comment by julienfalque** [comment]
I'm still in favor of *not* adding any option and ignore all `/** ... */` comments as soon as they contain an annotation by default. Maybe we could rename the fixer to something like `DocCommentToRegularCommentFixer` to avoid the PHPDoc rules?

Anyway we need to make a decision here /ping @keradus @SpacePossum

**Comment by masi** [comment]
> I'm still in favor of _not_ adding any option and ignore all `/** ... */` comments as soon as they contain an annotation by default.
Make this behaviour an opt-in and everybody should be happy.

> Maybe we could rename the fixer to something like `DocCommentToRegularCommentFixer` to avoid the PHPDoc rules?
Or have two rules.

**Comment by VincentLanglet** [comment]
I think this can be closed since the feature was introduced by https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5572

**Pull request closed without merging**